---
title: "Comparison not-filtered vs filtered"
author: "Paola Maragno, Alberto Pettenella"
date: "17-06-2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First we **load the list of significantly up regulated genes in each tissue with respect to both the other two** obtained as result of DE analysis starting from not-filtered data
```{r}
both_brain_nf <- read.table('both_brain_nf.txt', sep = ',')
both_liver_nf <- read.table('both_liver_nf.txt', sep = ',')
both_lung_nf  <- read.table('both_lung_nf.txt', sep = ',')
```

Then we **load the list of significantly up regulated genes in each tissue with respect to both the other two** obtained as result of DE analysis starting from filtered data
```{r}
both_brain_f <- read.table('BR_li_lu.txt', sep = ',')
both_liver_f <- read.table('LI_br_lu.txt', sep = ',')
both_lung_f  <- read.table('LU_li_br.txt', sep = ',')
```

Now we have to **extract the names of those genes that are filtered out since they are meaningless genes.** To do it we reload the data from GTEx portal
```{r, message=FALSE}
rse_brain <- recount3::create_rse_manual(
  project = "BRAIN",
  project_home = "data_sources/gtex",
  organism = "human",
  annotation = "refseq",
  type = "gene"
)

rse_liver <-recount3::create_rse_manual(
  project = "LIVER",
  project_home = "data_sources/gtex",
  organism = "human",
  annotation = "refseq",
  type = "gene"
)

rse_lung <- recount3::create_rse_manual(
  project = "LUNG",
  project_home = "data_sources/gtex",
  organism = "human",
  annotation = "refseq",
  type = "gene"
)
```

We **extract the three previously selected replicates for each tissue**
```{r}
rse_brain_selected <- rse_brain[,c(74,75,76)]
rse_liver_selected <- rse_liver[,c(74,75,78)]
rse_lung_selected <- rse_lung[,c(74,75,76)]
```

**This time we keep for each tissue only the rRNA genes, mitochondrial genes, pseudogenes or genes of unknown type**
```{r, message=FALSE}
library(edgeR)
library(recount3)
canonical <- paste("chr", seq(1,22), sep="")
canonical <- c(canonical, "chrX", "chrY")

filtered_genes_brain <- rse_brain_selected[
  # Ribosomal RNA
  rowData(rse_brain_selected)$gbkey == 'rRNA' |
    # Pseudogenes
    rowData(rse_brain_selected)$gbkey == 'Gene' |
    # Non-canonical Chromosomes and Mitochondrial DNA
    !rowRanges(rse_brain_selected)@seqnames %in% canonical |
    is.na(rowData(rse_brain_selected)$gbkey),
  ]

filtered_genes_liver <- rse_liver_selected[
  # Ribosomal RNA
  rowData(rse_liver_selected)$gbkey == 'rRNA' |
    # Pseudogenes
    rowData(rse_liver_selected)$gbkey == 'Gene' |
    # Non-canonical Chromosomes and Mitochondrial DNA
    !rowRanges(rse_liver_selected)@seqnames %in% canonical |
    # NAs
    is.na(rowData(rse_liver_selected)$gbkey),
  ]

filtered_genes_lung <- rse_lung_selected[
  # Ribosomal RNA
  rowData(rse_lung_selected)$gbkey == 'rRNA' |
    # Pseudogenes
    rowData(rse_lung_selected)$gbkey == 'Gene' |
    # Non-canonical Chromosomes and Mitochondrial DNA
    !rowRanges(rse_lung_selected)@seqnames %in% canonical |
    # NAs
    is.na(rowData(rse_lung_selected)$gbkey),
  ]
```

**We transform the coverage counts into raw counts, fish out the count tables and extract the gene names from it**
```{r}
assays(filtered_genes_brain)$counts <- transform_counts(filtered_genes_brain)
assays(filtered_genes_liver)$counts <- transform_counts(filtered_genes_liver)
assays(filtered_genes_lung)$counts <- transform_counts(filtered_genes_lung)

counts_filt_lung <- assays(filtered_genes_lung)$counts
counts_filt_brain <- assays(filtered_genes_brain)$counts
counts_filt_liver<- assays(filtered_genes_liver)$counts

rownames(counts_filt_lung) <- rowData(filtered_genes_lung)$gene_name
rownames(counts_filt_liver) <- rowData(filtered_genes_liver)$gene_name
rownames(counts_filt_brain) <- rowData(filtered_genes_brain)$gene_name

filt_name_brain <- as.vector(rownames(counts_filt_brain))
filt_name_liver <- as.vector(rownames(counts_filt_liver))
filt_name_lung <- as.vector(rownames(counts_filt_lung))
```


# Analysis of up regulated genes obtained only starting from not-filtered data

Extract the gene names of those genes that result up regulated in each tissue (vs both the other two) only when we don't filter the initial data
```{r}
only_nf_brain <- both_brain_nf[which(!both_brain_nf[,1] %in% both_brain_f[,1]),]
only_nf_liver <- both_liver_nf[which(!both_liver_nf[,1] %in% both_liver_f[,1]),]
only_nf_lung <- both_lung_nf[which(!both_lung_nf[,1] %in% both_lung_f[,1]),]
```

```{r}
length(only_nf_brain)
length(only_nf_liver)
length(only_nf_lung)
```

There are **589 brain genes, 434 liver genes and 402 lung genes that are up regulated in brain/liver/lung only in case of not filtering.**

For each tissue we compare these genes that have been found up regulated only when starting from not filtered data with the list of genes that we filtered out.
This is done to understand how many of those genes that result up regulated in a tissue - when we don't filter - are genes that can be filtered out since they are not interesting genes for a DE analysis
```{r}
length(only_nf_brain[which(!only_nf_brain %in% filt_name_brain)])
length(only_nf_liver[which(!only_nf_liver %in% filt_name_liver)])
length(only_nf_lung[which(!only_nf_lung %in% filt_name_lung)])
```

We can see that, among all the genes that are up regulated in each tissue only starting from not-filtered data, **only 76 brain genes, 1 liver gene and 64 lung genes are interesting DE genes (so no mitochondrial/rRNA/pseudogenes/unknown gene types).**

**This means that lot of genes that are returned as significantly up regulated in a tissue are not tissue-specific and so it is strongly advised to filter the data at the beginning of the analysis otherwise you have to do it on the final list of DE genes.**


# Functional enrichment analysis 

We can perform enrichment analysis of those genes - 76 brain genes, 1 liver gene and 64 lung genes - that are up regulated in each tissue with respect to both the other two only when we don't filter the data and that are genes of interest of that tissue.

```{r, message=FALSE, warning=FALSE}
library('enrichR')
setEnrichrSite("Enrichr")
websiteLive <- TRUE
```

## Brain

### Pathway enrichment analysis 
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_nf_brain[which(!only_nf_brain %in% filt_name_brain)], dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of BioPlanet 2019 database", enriched_pathway[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of KEGG 2021 Human database", enriched_pathway[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_nf_brain[which(!only_nf_brain %in% filt_name_brain)], dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Molecular Function 2021 database", enriched_ontologies[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_nf_brain[which(!only_nf_brain %in% filt_name_brain)], dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

## Liver

### Pathway enrichment analysis 
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_nf_liver[which(!only_nf_liver %in% filt_name_liver)], dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_nf_liver[which(!only_nf_liver %in% filt_name_liver)], dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_nf_liver[which(!only_nf_liver %in% filt_name_liver)], dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

## Lung

### Pathway enrichment analysis
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_nf_lung[which(!only_nf_lung %in% filt_name_lung)], dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of BioPlanet 2019 database", enriched_pathway[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of KEGG 2021 Human database", enriched_pathway[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_nf_lung[which(!only_nf_lung %in% filt_name_lung)], dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Molecular Function 2021 database", enriched_ontologies[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_nf_lung[which(!only_nf_lung %in% filt_name_lung)], dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```


#  Analysis of up regulated genes obtained only starting from filtered data

Extract the gene names of those genes that result up regulated in each tissue (vs both the other two) only when we filter the initial data
```{r}
only_f_brain <- both_brain_f[which(!both_brain_f[,1] %in% both_brain_nf[,1]),]
only_f_liver <- both_liver_f[which(!both_liver_f[,1] %in% both_liver_nf[,1]),]
only_f_lung <- both_lung_f[which(!both_lung_f[,1] %in% both_lung_nf[,1]),]
```

```{r}
length(only_f_brain)
length(only_f_liver)
length(only_f_lung)
```

There are **51 brain genes, 115 liver genes and 28 lung genes that are up regulated in brain/liver/lung only when starting from filtered data.**
These may be genes that are significantly differentially expressed in each tissue with respect to the other two that are hidden by the presence of lot of 'noise' genes and so don't result as DE when we don't filter the data (this is due to the fact that starting from a different number of genes the normalized counts can be different and as a consequence also the result of DE analysis may change).


# Functional enrichment analysis 

We can perform enrichment analysis of those genes - 51 brain genes, 115 liver gene and 28 lung genes - that are up regulated in each tissue with respect to both the other two only when we filter the data to understand if those genes that are lost when we don't filter the data are actually specific of that tissue and so important to detect as DE.

## Brain

### Pathway enrichment analysis 
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_f_brain, dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of BioPlanet 2019 database", enriched_pathway[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of KEGG 2021 Human database", enriched_pathway[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_f_brain, dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Molecular Function 2021 database", enriched_ontologies[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_f_brain, dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

## Liver

### Pathway enrichment analysis 
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_f_liver, dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of BioPlanet 2019 database", enriched_pathway[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of KEGG 2021 Human database", enriched_pathway[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_f_liver, dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Molecular Function 2021 database", enriched_ontologies[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_f_liver, dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

## Lung

### Pathway enrichment analysis
```{r, results="hide"}
dbs_pathway <- c("BioPlanet_2019", "WikiPathway_2021_Human", "KEGG_2021_Human")
if (websiteLive) {
    enriched_pathway <- enrichr(only_f_lung, dbs_pathway)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of BioPlanet 2019 database", enriched_pathway[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of WikiPathway 2021 Human database", enriched_pathway[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of KEGG 2021 Human database", enriched_pathway[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Ontologies enrichment analysis
```{r, results="hide"}
dbs_ontologies <- c("GO_Biological_Process_2021", "GO_Molecular_Function_2021", "GO_Cellular_Component_2021")
if (websiteLive) {
    enriched_ontologies <- enrichr(only_f_lung, dbs_ontologies)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of GO Biological Process 2021 database", enriched_ontologies[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Molecular Function 2021 database", enriched_ontologies[[2]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
if (websiteLive) plotEnrich(title = "Enriched terms of GO Cellular Component 2021 database", enriched_ontologies[[3]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```

### Cell types enrichment analysis
```{r, results="hide"}
dbs_celltypes <- c("Human_Gene_Atlas")
if (websiteLive) {
    enriched_celltypes <- enrichr(only_f_lung, dbs_celltypes)
}
```

```{r}
if (websiteLive) plotEnrich(title = "Enriched terms of Human Gene Atlas database", enriched_celltypes[[1]], showTerms = 5, numChar = 40, y = "Count", orderBy = "P.value")
```


In conclusion, from this analysis, we can say that **filtering the data is fundamental since lot of genes that result up regulated only starting from not-filtered data are indeed mitochondrial genes, ribosomal protein genes, psuedogenes or genes of unknown type of which we are not interested when performing DE analysis.**

In fact only 3.9% of the genes found to be up regulated in brain starting from not filtered data are not detected starting from filtered one, 0.07% in liver and 5.5% in lung.
From these numbers it is clear that **not-filtering the data doesn't give any advantage in terms of finding meaningful DE genes - with respect to starting from filtered data - and furthermore it requires to perform a filtering step on the final result of all the analysis.**  


